cuda/context_manager.cpp

```diff
@@ -0,0 +1,258 @@
+      stack.pop();
+    }
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no ctx stack for thread in ctx detroy exit callback");
+  }
+}
+
+
+// cuCtxSetCurrent_(entry|exit)
+void CUDAContextManager::ctx_set_current_entry(hpt_t hpt, CUcontext ctx) {
+  // std::cerr << "set current entry thread " << std::get<2>(hpt) << std::endl;
+  hpt_entry_ctx_[hpt] = ctx;
+}
+
+void CUDAContextManager::ctx_set_current_exit(hpt_t hpt, CUresult cuResult) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  // std::cerr << "set current exit thread " << std::get<2>(hpt) << std::endl;
+  CUcontext entry_ctx;
+  try {
+    entry_ctx = THAPI_AT(hpt_entry_ctx_, hpt);
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no entry context for thread in ctx set current exit callback");
+  }
+  try {
+    ensure_stack(hpt);
+    auto& stack = THAPI_AT(hpt_ctx_stack_, hpt);
+    if (!stack.empty()) {
+      stack.pop();
+    }
+    stack.push(entry_ctx);
+    // std::cerr << "stack size " << std::get<2>(hpt) << " " << hpt_ctx_stack_[hpt].size() << std::endl;
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no context stack for thread in ctx set current exit callback");
+  }
+}
+
+// cuCtxSetCurrent_(entry|exit)
+void CUDAContextManager::ctx_push_current_entry(hpt_t hpt, CUcontext ctx) {
+  hpt_entry_ctx_[hpt] = ctx;
+}
+
+void CUDAContextManager::ctx_push_current_exit(hpt_t hpt, CUresult cuResult) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  try {
+    ensure_stack(hpt);
+    auto entry_ctx = THAPI_AT(hpt_entry_ctx_, hpt);
+    hpt_ctx_stack_[hpt].push(entry_ctx);
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no entry context for thread in ctx push current exit callback");
+  }
+}
+
+
+// cuCtxPopCurrent_(entry|exit)
+void CUDAContextManager::ctx_pop_current_entry(hpt_t hpt) {
+  // noop
+}
+
+void CUDAContextManager::ctx_pop_current_exit(hpt_t hpt, CUresult cuResult,
+                                              CUcontext cuContext) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  // std::cerr << "pop current on thread " << std::get<2>(hpt) << std::endl;
+  hpt_ctx_stack_[hpt].pop();
+  // assert(popped_ctx == cuContext);
+}
+
+// cuStreamCreate_exit
+// Note: entry not needed
+void CUDAContextManager::stream_create_exit(hpt_t hpt, CUresult cuResult,
+                                            CUstream cuStream) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  CUcontext ctx;
+  bool found = get_top_context(hpt, &ctx);
+  if (!found) {
+    // TODO: how to we raise error here? Should not reach
+  }
+  hp_stream_t hp_stream_key = {std::get<0>(hpt), std::get<1>(hpt), cuStream}; 
+  
+  hp_stream_to_ctx_[hp_stream_key] = ctx;
+}
+
+void CUDAContextManager::ensure_stack(hpt_t hpt) {
+  if (hpt_ctx_stack_.count(hpt) == 0) {
+    hpt_ctx_stack_[hpt] = {};
+  }
```

Comment:
What is this function?
It sounds like a code smell to me. 

`{}` is the default value, look like it emulate the behavior of `[]`?

Reply:
Ya it was needed when using .at everywhere, which is still true for THAPI_AT which can be .at. But I think in this case I can use [] in the two places (ctx set and ctx create) where the stack can be created, and use AT elsewhere, and still have reasonable debugability.